src/include/mpir_pre.h

```diff
@@ -0,0 +1,31 @@
+/* -*- Mode: C; c-basic-offset:4 ; -*- */
+/*
+ * (C) 2018 by Argonne National Laboratory.
+ *     See COPYRIGHT in top-level directory.
+ */
+#ifndef MPIR_PRE_H_INCLUDED
+#define MPIR_PRE_H_INCLUDED
+
+/* N_BUILTIN definitions */
+
+/* This value should be set to greatest value used as the type index suffix in
+ * the predefined handles.  That is, look at the last two hex digits of all
+ * predefined datatype handles, take the greatest one, and convert it to decimal
+ * here. */
+/* FIXME: I will fix this by refactor the current datatype code out-of configure.ac */
+#define MPIR_DATATYPE_N_BUILTIN 71
```

Comment:
defining these values outside of the object header files seems a bit disconnected. is there some circular include we are avoiding?

Reply:
The reason I moved it to `mpir_pre.h` was because I want to use `MPIR_DATATYPE_N_PREDEFINED` and `MPIR_OP_N_BUILTIN` in `mpidpre.h` for `MPIDIG_win_info_args_t`. The current code for parsing "which_accumulate_ops" hint is too complex. It defines `MPIDIG_win_info_accu_op_shift_t ` and converts it manually to ops in `ch4r_win.c`, same complex code in OFI (including the weird magic numbers). I simplified the code in the PR by directly using the MPIR defined numbers.

I got compiling issues (basically header including circular) when including `mpir_datatype.h` in `mpidpre.h`.
```
../src/include/mpir_datatype.h: In function ‘MPIR_Datatype_free_contents’:
../src/include/mpir_datatype.h:293:10: error: ‘MPIR_Process’ undeclared (first use in this function)
      if (MPIR_Process.attr_free && datatype_ptr->attributes) {              \
          ^
../src/include/mpir_datatype.h:370:13: note: in expansion of macro ‘MPIR_Datatype_ptr_release’
             MPIR_Datatype_ptr_release(old_dtp);
```
That is why I think it might make more sense to put the number definitions to `mpir_pre.h` so it is clean to include them in all other headers.

Reply:
Is that part of this PR? I do see `MPIR_OP_N_BUILTIN` used in `mpidpre.h` in these changes, but nothing relating to datatype builtin or predefined constants. `mpir_op.h` is not device dependent at this time, therefore is included before `mpidpre.h` and causes no compilation issue.

Reply:
Good point. I checked again. No need to include "mpir_datatype.h" for this refactoring PR. I will need include it only for the ongoing one which includes datatype information into the atomics hint. Let me move the `mpir_pre.h` commit to the ongoing PR.

Reply:
I removed the commit of `mpir_pre` in this PR. But I definitely need resolve the header inclusion problem in my ongoing PR since I will use `MPIR_DATATYPE_N_PREDEFINED` in `mpidpre.h`. Do you have any better idea to address the issue?

Reply:
I think in this time crunch it would be OK to just redefine it (either `#ifndef` or some new `MPIDI_` name) and add a FIXME comment.

Reply:
Alright. Will do it in the other ongoing PR which uses datatypes. Btw, this PR has been cleaned up to address all comments.